File: src/main/org/nlogo/window/SpeedSliderPanel.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window


import java.awt.{ BorderLayout, Color, Dimension, Graphics, GridBagConstraints, GridBagLayout }
import java.awt.event.{ MouseAdapter, MouseEvent, MouseWheelEvent, MouseWheelListener }
import javax.swing.{ JLabel, JPanel, JSlider }
import javax.swing.event.{ ChangeEvent, ChangeListener }
import org.nlogo.api.{ I18N, Version }
import org.nlogo.awt.Fonts
import org.nlogo.log.Logger

class SpeedSliderPanel(workspace: GUIWorkspace, labelsBelow: Boolean)
    extends JPanel with ChangeListener with Events.LoadBeginEventHandler {
  implicit val i18nName = I18N.Prefix("tabs.run.speedslider")
  private val normal = new SpeedLabel(I18N.gui("normalspeed"))
  private val speedSlider = new SpeedSlider(workspace.speedSliderPosition.toInt)
  speedSlider.setFocusable(false)
  speedSlider.addChangeListener(this)
  speedSlider.addMouseListener(new MouseAdapter {
    // when we release the mouse if it's kinda close to the
    // center snappy snap.  ev 2/22/07
    override def mouseReleased(e: MouseEvent) = {
        val value = speedSlider.getValue
        if(value >= -10 && value <= 10)
        speedSlider.setValue(0)
    }
  })
  speedSlider.setOpaque(false)
  Fonts.adjustDefaultFont(normal)
  setOpaque(false)
  if(labelsBelow) {
    val gridbag = new GridBagLayout
    val c = new GridBagConstraints
    setLayout(gridbag)
    c.gridwidth = GridBagConstraints.REMAINDER
    add(speedSlider, c)
    c.gridwidth = 1
    c.anchor = GridBagConstraints.CENTER
    add(normal, c)
  } else {
    val layout = new BorderLayout
    layout.setVgap(0)
    setLayout(layout)
    add(speedSlider, BorderLayout.CENTER)
    add(normal, BorderLayout.EAST)
  }
  enableLabels(0)

  override def setEnabled(enabled: Boolean) = {
    speedSlider.setEnabled(enabled)
    // if we do setVisible() on the label, that changes the layout
    // which is a bit jarring, so do this instead - ST 9/16/08
    if(enabled) stateChanged(null) else normal.setText(" ")
  }

  def stateChanged(e: ChangeEvent) = {
    var value = speedSlider.getValue
    // adjust the speed reported to the workspace
    // so there isn't a big gap between the snap area
    // and outside the snap area. ev 2/22/07
    if(value < -10)     value += 10
    else if(value > 10) value -= 10
    else                value = 0
    workspace.speedSliderPosition(value / 2)
    if(Version.isLoggingEnabled)
      Logger.logSpeedSlider(value)
    enableLabels(value)
    workspace.updateManager.nudgeSleeper()
  }

  def enableLabels(value: Int) = normal.setText(
    if(value == 0)
      (if(labelsBelow) " "*6 else "") + I18N.gui("normalspeed")
    else if(value < 0)
      I18N.gui("slower") + (if(labelsBelow) " "*25 else "")
    else
      (if(labelsBelow) " "*25 else "") + I18N.gui("faster"))

  def handle(e: Events.LoadBeginEvent) = speedSlider.reset()

  def setValue(speed: Int) = {
    speedSlider.setValue(speed)
    enableLabels(workspace.speedSliderPosition.toInt)
  }
  def getValue = speedSlider.getValue
  def getMaximum = speedSlider.getMaximum
  override def isEnabled = speedSlider.isEnabled

  private class SpeedSlider(defaultSpeed: Int)
      extends JSlider(-110, 112, defaultSpeed) with MouseWheelListener {
    setExtent(1)
    setToolTipText("Adjust speed of model")
    addMouseWheelListener(this)

    override def getPreferredSize = new Dimension(180, super.getPreferredSize.height)
    override def getMinimumSize = new Dimension(60, super.getPreferredSize.height)
    
    def reset() = setValue(0)

    def mouseWheelMoved(e: MouseWheelEvent) = setValue(getValue - e.getWheelRotation)

    override def paint(g: Graphics) = {
      val bounds = getBounds()
      val x = bounds.x + (bounds.width / 2) - 1
      g.setColor(Color.gray)
      g.drawLine(x, bounds.y + (bounds.height * 3 / 4), x, bounds.y + bounds.height)
      super.paint(g)
    }
  }

  private class SpeedLabel(label: String) extends JLabel(label) {
    override def getPreferredSize = getMinimumSize
    override def getMinimumSize = {
      val d = super.getMinimumSize
      val fontMetrics = getFontMetrics(getFont)
      d.width = StrictMath.max(d.width,
        fontMetrics.stringWidth(if(!labelsBelow) I18N.gui("normalspeed") else " "*25 + I18N.gui("faster") + 10))
      d
    }
  }
}

File: src/main/org/nlogo/window/SpeedSliderPanel.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.api.I18N;

import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;

public strictfp class SpeedSliderPanel
    extends javax.swing.JPanel
    implements
    java.awt.event.MouseListener,
    javax.swing.event.ChangeListener,
    org.nlogo.window.Events.LoadBeginEvent.Handler {
  private final GUIWorkspace workspace;

  final SpeedSlider speedSlider;

  final javax.swing.JLabel normal = new SpeedLabel(I18N.guiJ().get("tabs.run.speedslider.normalspeed"));

  private final boolean labelsBelow;

  public SpeedSliderPanel(GUIWorkspace workspace, boolean labelsBelow) {
    this.workspace = workspace;
    this.labelsBelow = labelsBelow;
    speedSlider = new SpeedSlider((int) workspace.speedSliderPosition());
    speedSlider.setFocusable(false);
    speedSlider.addChangeListener(this);
    speedSlider.addMouseListener(this);
    speedSlider.setOpaque(false);
    org.nlogo.awt.Fonts.adjustDefaultFont(normal);
    setOpaque(false);
    if (labelsBelow) {
      GridBagLayout gridbag = new GridBagLayout();
      GridBagConstraints c = new GridBagConstraints();
      setLayout(gridbag);
      c.gridwidth = GridBagConstraints.REMAINDER;
      add(speedSlider, c);
      c.gridwidth = 1;
      c.anchor = GridBagConstraints.CENTER;
      add(normal, c);
    } else {
      java.awt.BorderLayout layout = new java.awt.BorderLayout();
      layout.setVgap(0);
      setLayout(layout);
      add(speedSlider, java.awt.BorderLayout.CENTER);
      add(normal, java.awt.BorderLayout.EAST);
    }
    enableLabels(0);
  }

  @Override
  public void setEnabled(boolean enabled) {
    speedSlider.setEnabled(enabled);
    // if we do setVisible() on the label, that changes the layout
    // which is a bit jarring, so do this instead - ST 9/16/08
    if (enabled) {
      stateChanged(null);
    } else {
      normal.setText(" ");
    }
  }

  public void stateChanged(javax.swing.event.ChangeEvent e) {
    int value = speedSlider.getValue();
    // adjust the speed reported to the workspace
    // so there isn't a big gap between the snap area
    // and outside the snap area. ev 2/22/07
    if (value < -10) {
      value += 10;
    } else if (value > 10) {
      value -= 10;
    } else {
      value = 0;
    }
    workspace.speedSliderPosition(value / 2);
    if (org.nlogo.api.Version.isLoggingEnabled()) {
      org.nlogo.log.Logger.logSpeedSlider(value);
    }
    enableLabels(value);
    workspace.updateManager().nudgeSleeper();
  }

  void enableLabels(int value) {
    if (value == 0) {
      if (labelsBelow) {
        normal.setText("        " + I18N.guiJ().get("tabs.run.speedslider.normalspeed"));
      } else {
        normal.setText(I18N.guiJ().get("tabs.run.speedslider.normalspeed"));
      }
    } else if (value < 0) {
      if (labelsBelow) {
        normal.setText(I18N.guiJ().get("tabs.run.speedslider.slower") + "                         ");
      } else {
        normal.setText(I18N.guiJ().get("tabs.run.speedslider.slower"));
      }
    } else {
      if (labelsBelow) {
        normal.setText("                         " + I18N.guiJ().get("tabs.run.speedslider.faster"));
      } else {
        normal.setText(I18N.guiJ().get("tabs.run.speedslider.faster"));
      }
    }
  }

  // mouse listener junk
  public void mouseClicked(java.awt.event.MouseEvent e) {
  }

  public void mousePressed(java.awt.event.MouseEvent e) {
  }

  public void mouseEntered(java.awt.event.MouseEvent e) {
  }

  public void mouseExited(java.awt.event.MouseEvent e) {
  }

  // when we release the mouse if it's kinda close to the
  // center snappy snap.  ev 2/22/07
  public void mouseReleased(java.awt.event.MouseEvent e) {
    int value = speedSlider.getValue();
    if ((value <= 10 && value > 0) ||
        (value >= -10 && value < 0)) {
      speedSlider.setValue(0);
    }
  }

  public void handle(org.nlogo.window.Events.LoadBeginEvent e) {
    speedSlider.reset();
  }

  public void setValue(int speed) {
    if (speedSlider.getValue() != speed) {
      speedSlider.setValue(speed);
    }
    enableLabels((int) workspace.speedSliderPosition());
  }

  public int getValue() {
    return speedSlider.getValue();
  }

  public int getMaximum() {
    return speedSlider.getMaximum();
  }

  @Override
  public boolean isEnabled() {
    return speedSlider.isEnabled();
  }

  private class SpeedSlider
      extends javax.swing.JSlider
      implements java.awt.event.MouseWheelListener {
    SpeedSlider(int defaultSpeed) {
      super(-110, 112, defaultSpeed);
      setExtent(1);
      setToolTipText("Adjust speed of model");
      addMouseWheelListener(this);
    }

    @Override
    public java.awt.Dimension getPreferredSize() {
      return new java.awt.Dimension
          (180, super.getPreferredSize().height);
    }

    @Override
    public java.awt.Dimension getMinimumSize() {
      return new java.awt.Dimension
          (60, super.getPreferredSize().height);
    }

    void reset() {
      setValue(0);
    }

    public void mouseWheelMoved(java.awt.event.MouseWheelEvent e) {
      setValue(getValue() - e.getWheelRotation());
    }

    @Override
    public void paint(java.awt.Graphics g) {
      java.awt.Rectangle bounds = getBounds();
      int x = bounds.x + (bounds.width / 2) - 1;
      g.setColor(java.awt.Color.gray);
      g.drawLine(x, bounds.y + (bounds.height * 3 / 4), x, bounds.y + bounds.height);
      super.paint(g);
    }
  }

  private class SpeedLabel
      extends javax.swing.JLabel {
    SpeedLabel(String label) {
      super(label);
    }

    @Override
    public java.awt.Dimension getPreferredSize() {
      return getMinimumSize();
    }

    @Override
    public java.awt.Dimension getMinimumSize() {
      java.awt.Dimension d = super.getMinimumSize();
      java.awt.FontMetrics fontMetrics = getFontMetrics(getFont());
      if (!labelsBelow) {
        d.width = StrictMath.max(d.width,
            fontMetrics.stringWidth(I18N.guiJ().get("tabs.run.speedslider.normalspeed")));
      } else {
        d.width = StrictMath.max(d.width,
            fontMetrics.stringWidth
                ("                         " + I18N.guiJ().get("tabs.run.speedslider.faster")) + 10);

      }
      return d;
    }
  }
}
